src/shared/serialize.c

```diff
@@ -79,7 +80,23 @@ int serialize_item_format(FILE *f, const char *key, const char *format, ...) {
         return 1;
 }
 
-int serialize_fd(FILE *f, FDSet *fds, const char *key, int fd) {
+int serialize_prepare_fd_array_or_set(int fd, FDSet *fds, int **fds_array, size_t *n_fds_array) {
+        assert(fd >= 0);
+        assert(fds || (fds_array && n_fds_array));
+
+        if (fds)
+                return fdset_put_dup(fds, fd);
+
+        *fds_array = GREEDY_REALLOC(*fds_array, *n_fds_array + 1);
```

Comment:
GREEDY_REALLOC() already updates the first param anyway, no need to assign it's return value